drivers/i2c/i2c_wch.c

```diff
@@ -0,0 +1,498 @@
+		wch_i2c_handle_btf(dev);
+	} else if ((status & I2C_STAR1_TXE) && write) {
+		wch_i2c_handle_txe(dev);
+	} else if ((status & I2C_STAR1_RXNE) && (!write)) {
+		wch_i2c_handle_rxne(dev);
+	}
+}
+
+static void i2c_wch_error_isr(const struct device *dev)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+	uint16_t status = regs->STAR1;
+
+	if (status & (I2C_STAR1_AF | I2C_STAR1_ARLO | I2C_STAR1_BERR)) {
+		if (status & I2C_STAR1_AF) {
+			regs->CTLR1 |= I2C_CTLR1_STOP;
+		}
+
+		data->current.err |= ((status >> 8) & 0x7);
+
+		regs->STAR1 &= ~(I2C_STAR1_AF | I2C_STAR1_ARLO | I2C_STAR1_BERR);
+
+		atomic_set(&data->xfer_done, 1);
+	}
+}
+
+static void wch_i2c_msg_init(const struct device *dev, struct i2c_msg *msg, uint16_t addr)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+
+	atomic_set(&data->xfer_done, 0);
+
+	data->current.msg = msg;
+	data->current.idx = 0U;
+	data->current.err = 0U;
+	data->current.addr = addr;
+
+	regs->CTLR1 |= I2C_CTLR1_PE;
+	regs->CTLR1 |= I2C_CTLR1_ACK;
+
+	if (msg->flags & I2C_MSG_RESTART) {
+		if (regs->CTLR1 & I2C_CTLR1_STOP) {
+			regs->CTLR1 &= ~I2C_CTLR1_STOP;
+		}
+		regs->CTLR1 |= I2C_CTLR1_START;
+	}
+}
+
+static int32_t wch_i2c_msg_end(const struct device *dev)
+{
+	struct i2c_wch_data *data = dev->data;
+
+	if (!data->current.err) {
+		return 0;
+	}
+
+	if (data->current.err & (I2C_STAR1_ARLO >> 8)) {
+		LOG_DBG("ARLO");
+	}
+
+	if (data->current.err & (I2C_STAR1_AF >> 8)) {
+		LOG_DBG("NACK");
+	}
+
+	if (data->current.err & (I2C_STAR1_BERR >> 8)) {
+		LOG_DBG("ERR");
+	}
+
+	data->current.err = 0;
+
+	return -EIO;
+}
+
+static void wch_i2c_config_interrupts(I2C_TypeDef *regs, bool enable)
+{
+	if (enable) {
+		regs->CTLR2 |= (I2C_CTLR2_ITERREN | I2C_CTLR2_ITEVTEN | I2C_CTLR2_ITBUFEN);
+	} else {
+		regs->CTLR2 &= ~(I2C_CTLR2_ITERREN | I2C_CTLR2_ITEVTEN | I2C_CTLR2_ITBUFEN);
+	}
+}
+
+static int32_t wch_i2c_begin_transfer(const struct device *dev, struct i2c_msg *msg, uint16_t addr)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+
+	wch_i2c_msg_init(dev, msg, addr);
+
+	wch_i2c_config_interrupts(regs, true);
+
+	k_sched_lock();
+	while (atomic_get(&data->xfer_done) == 0) {
+	}
+	k_sched_unlock();
+
+	return wch_i2c_msg_end(dev);
+}
+
+static void wch_i2c_finish_transfer(const struct device *dev)
+{
+	const struct i2c_wch_config *config = dev->config;
+	I2C_TypeDef *regs = config->regs;
+
+	wch_i2c_config_interrupts(regs, false);
+
+	regs->CTLR1 &= ~I2C_CTLR1_PE;
+}
+
+static void wch_i2c_configure_timing(I2C_TypeDef *regs, uint32_t clock_rate, uint32_t bitrate)
+{
+	uint16_t clock_config;
+	uint16_t freq_range;
+
+	if (bitrate <= I2C_BITRATE_STANDARD) {
+		clock_config = MAX((uint16_t)(clock_rate / (bitrate << 1)), 4);
```

Comment:
nit: for consistency with the (bitrate * 3) below, use (bitrate * 2) here.

Reply:
Done